File: translated/f-to-c/DRB064-outeronly2-orig-no.c
/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Copyright (c) 2017-20, Lawrence Livermore National Security, LLC
and DataRaceBench project contributors. See the DataRaceBench/COPYRIGHT file for details.

SPDX-License-Identifier: (BSD-3-Clause)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*/

//Only the outmost loop can be parallelized.
//The inner loop has loop carried true data dependence.
//However, the loop is not parallelized so no race condition.

#include <omp.h>
#include <stdio.h>
#include <stdlib.h>

void foo()
{
    int i, j, n, m, len;
    float** b;

    len = 100;

    // Allocate 2D array
    b = (float**)malloc(len * sizeof(float*));
    for (i = 0; i < len; i++) {
        b[i] = (float*)malloc(len * sizeof(float));
    }

    n = len;
    m = len;

    #pragma omp parallel for private(j)
    for (i = 0; i < n; i++) {
        for (j = 1; j < m; j++) {
            b[i][j] = b[i][j-1];
        }
    }

    // Free 2D array
    for (i = 0; i < len; i++) {
        free(b[i]);
    }
    free(b);
}

int main()
{
    foo();
    return 0;
} 

File: translated/f-to-go/DRB064-outeronly2-orig-no.go
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// Copyright (c) 2017-20, Lawrence Livermore National Security, LLC
// and DataRaceBench project contributors. See the DataRaceBench/COPYRIGHT file for details.
//
// SPDX-License-Identifier: (BSD-3-Clause)
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

//Only the outmost loop can be parallelized.
//The inner loop has loop carried true data dependence.
//However, the loop is not parallelized so no race condition.

package main

import (
	"runtime"
	"sync"
)

func foo() {
	var n, m, length int
	var b [][]float32

	length = 100
	b = make([][]float32, length)
	for i := 0; i < length; i++ {
		b[i] = make([]float32, length)
	}
	n = length
	m = length

	//$omp parallel do private(j)
	var wg sync.WaitGroup
	numCPU := runtime.NumCPU()
	chunkSize := n / numCPU
	if chunkSize < 1 {
		chunkSize = 1
	}

	for start := 1; start <= n; start += chunkSize {
		end := start + chunkSize - 1
		if end > n {
			end = n
		}
		wg.Add(1)
		go func(start, end int) {
			defer wg.Done()
			for i := start; i <= end; i++ {
				for j := 2; j <= m; j++ { // j is private, inner loop has dependence but is not parallelized
					b[i-1][j-1] = b[i-1][j-2] // No race - proper data partitioning
				}
			}
		}(start, end)
	}
	wg.Wait()
	//$omp end parallel do
}

func main() {
	foo()
}
